### ISSUE TITLE
Openidc.introspect escape username and password when using client_secret_basic authentication 

### ISSUE THREAD

**Issue by happycujo** [zmartzone/lua-resty-openidc#247]
We want to validate a access token to a WSO2 endpoint, which require a username(**client_id**) and password (**client_secret**). When use the open resty **openidc.introspect**  function to validate the access token using the  **client_secret_basic** authentication mode. The username contains an invalid character for an uri (@) as it is an email address.  When looking at the code, we noticed that introspect call internally the **call_token_endpoint** function. This function  convert the username (client_id) invalid character (@ become %40) using **ngx.escape_uri** prior sending the http request. 

--- Line 413 of openidc.lua
`  headers.Authorization = "Basic " .. b64(ngx.escape_uri(opts.client_id) .. ":" ..ngx.escape_uri(opts.client_secret))`

The WSO2 server receive the altered username and password, and refuse the introspection request (401 HTTP code) as the username and password are not matching its data. 

By enabling the debug mode on the server, we intercepted the request. There are no error associated in the log
```
POST /oauth2/introspect HTTP/1.1
User-Agent: lua-resty-http/0.13 (Lua) ngx_lua/10013
Content-Length: 1444
Content-Type: application/x-www-form-urlencoded
Authorization: Basic <base 64 username:password <--ngx.escape_uri applied>
Host: <Hostname>
```

Performing the same request with postman (an API test software) with the _unaltered_ username and password (still encoded in base64) showed us that the WSO2 endpoint worked has expected. 

As a side note, the **client_secret_post** authentication method provide the client_id and client_secret in the body unaltered unlike the client_secret_basic. Unfortunately for us,  WSO2 is not supporting this way to provide username and password to their introspection End Point.

###### Environment

- lua-resty-openidc version (e.g. 1.7.1)
- WSO2 5.7.0

###### Expected behaviour
The **openidc. introspection** function should not alter the content of the client_id and client_secret as it is already encoded once in base64. 

The line  
`headers.Authorization = "Basic " .. b64(ngx.escape_uri(opts.client_id) .. ":" .. ngx.escape_uri(opts.client_secret))`

could be replaced with
`headers.Authorization = "Basic " .. b64(opts.client_id .. ":" .. opts.client_secret)`

 If there some use case where this behavior is still required, an option could be provided to enable or disable it.

###### Actual behaviour
The **openidc. introspection** function replace the invalid uri character (@ in my case) and WSO2 can not authenticate the user (HTTP 401)

###### Minimized example
1. Setup a WSO2 server a create a account using an email address for the username
2. Add the code below to your openresty nginx.conf file prior to read the default location
3. Make sure to edit server host, username and password to match your configuration

###### Configuration and NGINX server log files
			local opts = {
				introspection_endpoint="https://server:443/oauth2/introspect",
			    client_id = "email@wso2.com",
			    client_secret = "password",
				ssl_verify = "no",
				introspection_expiry_claim = "exp",
				introspection_endpoint_auth_method = "client_secret_basic"
			} 

			-- call introspect for OAuth 2.0 Bearer Access Token validation
			local res, err = require("resty.openidc").introspect(opts)

			if err then
				ngx.status = 403
				ngx.say(err)
				ngx.exit(ngx.HTTP_FORBIDDEN)
			end

**Comment by zandbelt** [comment]
The URL encoding was explicitly added since the spec requires it, see:
https://github.com/zmartzone/lua-resty-openidc/issues/204

**Comment by zandbelt** [comment]
in your case it is probably best to use a URL-safe secret

**Comment by happycujo** [comment]
Thanks for the quick turn around! It seems that it is WSO2 that is not respecting OpenID standard (or at least the version used in our organisation). :(

I had tried your suggestion to provide the username and password in the url.  However i stumbled on an another issue. Using an url such as https://email@wso2.com:password@server.com/oauth2/introspect seems to confuse the openidc.introspect function. He was not able to parse the host correctly and returned the wso2.com instead of the server.com from the url. I tried a couple of variant using %40 to replace the @ but ended up with the same result.
